### ISSUE TITLE
Syntax sugar: allow trailing commas in bracket-enclosed lists

### ISSUE THREAD

**Issue by fsackur** [PowerShell/PowerShell#16844]
### Summary of the new feature / enhancement

I'm keen on #8873, but why stop at method calls and param blocks?

Python allows this syntax:

``` python
list = [
    'foo',
    'bar',
]
```

The trailing comma after `'bar'` is what I would like to see in PowerShell.

The advantages are:

- appending to a list is a one-line diff, not a 2-line diff
- commenting out the last list entry is not a syntax error

I would like all bracket-enclosed list expressions to support this:

``` powershell
$List = (
   'foo',
   'bar',
)

$List = @(
   'foo',
   'bar',
)
```

I do NOT propose support without the brackets, because it seems like a serious breaking change:

``` powershell
# This should remain a syntax error
$List = 'foo', bar',
$a = 123
```

**Comment by zett42** [comment]
Array literals already allow you to remove the comma, so you get one-line diffs:
~~~powershell
# This is already valid syntax
$List = @(
   'foo'
   'bar'
)
~~~

**Comment by SeeminglyScience** [comment]
> Array literals already allow you to remove the comma, so you get one-line diffs:

Note that's not really an array literal at that point. It's a bit closer to:

```powershell
$List = & {
    'foo'
    'bar'
}
```

And will only work with `@()`, not `()`.

Hashtable key value pairs cannot be separated with `,`, they can only be separated with `;` or new lines.

**Comment by iSazonov** [comment]
> Hashtable key value pairs cannot be separated with `,`, they can only be separated with `;` or new lines.

And tricky question - why is there `;`? 😸

**Comment by SeeminglyScience** [comment]
> And tricky question - why is there `;`? 😸

Probably more incidental than anything. It's just the statement terminator

**Comment by zett42** [comment]
> Note that's not really an array literal at that point.

Thanks for correcting. It's called the [array sub-expression operator](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_arrays?view=powershell-7.2#the-array-sub-expression-operator).

Are there any practical differences compared to real array literals, when defining _arrays over multiple lines_? I. e. which one of the following  would be preferrable under different circumstances?

~~~powershell
# Wishful thinking
$List = 
   'foo',
   'bar',

# Wishful thinking
$List = (
   'foo',
   'bar',
)

# Already possible
$List = @(
   'foo'
   'bar'
)
~~~

The 1st example silently changes behaviour when we miss the comma after any element that is not the last one. It is also less readable in my opinion, because it doesn't visually group the array elements, aside from the indentation.

If we miss comma in the 2nd example, we get syntax error "Missing closing ')' in expression", which is better.

I still prefer the last one as there is no comma to be missed and it looks visually cleaner and more similar to how we define multiline hashtable literals `@{ }`.

Is there a possible performance benefit of `( )` over `@( )`?

**Comment by SeeminglyScience** [comment]
> Is there a possible performance benefit of `( )` over `@( )`?

It's more the commas than anything. `('one', 'two')` and `@('one', 'two')` compile identically. But using new lines to separate makes a significant difference in some scenarios.

<details>

<summary>

## Click for compilation

```powershell
{ $list = @(
'w',
'b'
) } | ScriptBlockDisassembler\Get-ScriptBlockDisassembly -Minimal
```

</summary>

```csharp
// ScriptBlock.EndBlock
try
{
    locals.Item009 = (object)new object[] { (object)"w", (object)"b" };
}
catch (FlowControlException)
{
    throw;
}
catch (Exception exception)
{
    ExceptionHandlingOps.CheckActionPreference(funcContext, exception);
}
```

</details>

<details>

<summary>

## Click for compilation

```powershell
{ $list = @(
'w'
'b'
) } | ScriptBlockDisassembler\Get-ScriptBlockDisassembly -Minimal
```

</summary>

```csharp
// ScriptBlock.EndBlock
try
{
    locals.Item009 = (object)Fake.Dynamic<Func<CallSite, List<object>, object[]>>(PSToObjectArrayBinder.Get())(
        try
        {
            Pipe oldPipe0 = funcContext._outputPipe;
            List<object> resultList1 = new List<object>();
            funcContext._outputPipe = new Pipe(resultList1);
            oldPipe0.SetVariableListForTemporaryPipe(funcContext._outputPipe);
            int stmt = 0;

            unnamed_label_0:
            try
            {
                switch (stmt)
                {
                    case 0:
                        goto unnamed_label_1;
                        break;

                    case 1:
                        goto unnamed_label_2;
                        break;

                    case 2:
                        goto unnamed_label_3;
                        break;
                }

                unnamed_label_1:
                stmt = 1;
                funcContext._outputPipe.Add("w");

                unnamed_label_2:
                stmt = 2;
                funcContext._outputPipe.Add("b");
            }
            catch (FlowControlException)
            {
                throw;
            }
            catch (Exception exception)
            {
                ExceptionHandlingOps.CheckActionPreference(funcContext, exception);
                goto unnamed_label_0;
            }

            unnamed_label_3:

            return resultList1;
        }
        finally
        {
            return funcContext._outputPipe = oldPipe0;
        });
}
catch (FlowControlException)
{
    throw;
}
catch (Exception exception)
{
    ExceptionHandlingOps.CheckActionPreference(funcContext, exception);
}
```

</details>

TL;DR: The former just makes an array, exactly like you'd expect. The latter spins up an output pipe and evaluates every line as a statement, retrieves the result from the output pipe and then converts to an array.

Granted in the majority of powershell scripts none of this matters. If however you are allocating a lot of arrays in a loop, the difference may be significant at higher scales.

**Comment by scriptingstudio** [comment]
> If however you are allocating a lot of arrays in a loop, the difference may be significant at higher scales.

Is it also true for hashtables? They are some kind of arrays.

**Comment by SeeminglyScience** [comment]
> Is it also true for hashtables? They are some kind of arrays.

If you mean the difference between these:

```powershell
@{
    key = 'value'
    key2 = 'value'
}

# And

@{
    key = 'value';
    key2 = 'value';
}
```

No, there's no difference.

**Comment by fsackur** [comment]
I see https://github.com/PowerShell/PowerShell/pull/16851#issuecomment-1034299246

The same reasoning would apply here. I believe this enhancement would be closed too for the same reasons, so I'll close it now.
